src/subgrid.c

```diff
@@ -183,10 +189,33 @@ int subgrid_update(colloids_info_t * cinfo, hydro_t * hydro) {
 
           if (p_colloid->s.type != COLLOID_TYPE_SUBGRID) continue;
 
-	  drag = reta*(1.0/p_colloid->s.a0 - 1.0/p_colloid->s.ah);
+	  drag = reta*(1.0/p_colloid->s.a0 - 1.0/p_colloid->s.al);
+
+	  if (noise_flag == 0) {
+	    frand[X] = 0.0; frand[Y] = 0.0; frand[Z] = 0.0;
+	  }
+	  else {
+	    for (ia = 0; ia < 3; ia++) {
+	      while (1) {
+		/* To keep the random correction smaller than 3 sigma.
+		 * Otherwise, a large thermal fluctuation may cause a
+		 * numerical problem. */
+		util_ranlcg_reap_gaussian(&p_colloid->s.rng, ran);
+		if (fabs(ran[0]) < 3.0) {
+		  frand[ia] = sqrt(2.0*kt*drag)*ran[0];
+		  break;
+		}
+		if (fabs(ran[1]) < 3.0) {
+		  frand[ia] = sqrt(2.0*kt*drag)*ran[1];
+		  break;
+		}
+	      }
+	    }
+	  }
 
 	  for (ia = 0; ia < 3; ia++) {
-	    p_colloid->s.v[ia] = p_colloid->fsub[ia] + drag*p_colloid->fex[ia];
+	    p_colloid->s.v[ia] = p_colloid->fsub[ia] + drag*p_colloid->fex[ia]
+                               + frand[ia];
```

Comment:
int v2=p_colloid->s.v[X]*p_colloid->s.v[X]+p_colloid->s.v[Y]*p_colloid->s.v[Y]+p_colloid->s.v[Z]*p_colloid->s.v[Z];
if(v2>=0.25) {
       p_colloid->s.v[X]*=0.5;
       p_colloid->s.v[Y]*=0.5;
       p_colloid->s.v[Z]*=0.5;
       printf("Large velocity correction. vel idx: %d; v: %lf,%lf,%lf;\n",p_colloid->s.index,p_colloid->s.v[0],p_colloid->s.v[1],p_colloid->s.v[2]);
}
/* To avoid too large velocity. It is valid only if v2 very rarely exceeds the threshold. It is equivalent to applying a large random force to the particle. */
/* Hi Kevin, for a system with a lot of polymers, very occasionally, the code will crash because the velocity is too large. This happens because of the resonance of a large external force and a large random force. */